Support array filters

memicmiralem opened the pull request:
Hello there

I was using this package and noticed that prisma array filters are not supported so I tried to do it myself.

These code changes enable use of following:
- https://www.prisma.io/docs/orm/reference/prisma-client-reference#array_contains
- https://www.prisma.io/docs/orm/reference/prisma-client-reference#array_starts_with
- https://www.prisma.io/docs/orm/reference/prisma-client-reference#array_ends_with

I wrote tests and it seems fine, let me know if I missed something.

memicmiralem committed:
add array operation types

memicmiralem committed:
support parsing array filters

memicmiralem committed:
add tests for array filters

Valerionn committed:
Add prettier, remove pnpm files

Valerionn committed:
Merge branch 'main' into feature/support-array-filters

# Conflicts:
#	prisma-filter/src/filter.parser.ts
#	prisma-filter/src/prisma.type.ts

Valerionn committed:
Add prettier to tests as well

Valerionn committed:
Merge branch 'main' into feature/support-array-filters

Valerionn commented on prisma-filter/tests/filter.type.test.ts:
```diff
@@ -156,3 +174,104 @@ test('ISearch', () => {
+    array_contains: ['val1', 'val2', 'val3'],
+  });
+
+  findOptions = generateArrayContainsQuery([1, 2, 3]);
```
Unfortunately, REST parameters are always parsed as strings. Therefore, to test this, we have to use strings, e.g.

```ts
findOptions = generateArrayContainsQuery(['1', '2', '3']);
// ...
findOptions = generateArrayContainsQuery('123');
```

This library has to then correctly convert values, as described in my other comment.

Valerionn commented on prisma-filter/src/filter.parser.ts:
```diff
@@ -104,9 +108,17 @@ export class FilterParser<TDto, TWhereInput> {
+        type === FilterOperationType.InStrings ||
+        type === FilterOperationType.NotInStrings ||
+        type === FilterOperationType.ArrayContains
+      )
         return rawValue;
-      if (type !== FilterOperationType.In && type !== FilterOperationType.NotIn) {
+      if (
+        type !== FilterOperationType.In &&
+        type !== FilterOperationType.NotIn //&&
+        // type !== FilterOperationType.ArrayContains
+      ) {
```
Since the input values here are always strings (and Prisma is usually very strict with their typings, and you can't just pass a string for a number array), I think the best way is to offer two filter operations: `ArrayContains` and `ArrayContainsStrings`. For `ArrayContainsStrings`, you can directly return the raw value - and `ArrayContains` then correctly casts the values to numbers (you only have to include your commented-out code again, and it should work).

boolean arrays are currently not truly supported - but that's a library problem that I can fix myself if you don't want to add a bool conversion here.

Valerionn commented on prisma-filter/src/filter.parser.ts:
```diff
@@ -120,6 +132,10 @@ export class FilterParser<TDto, TWhereInput> {
     }
 
+    if (type === FilterOperationType.ArrayContains) {
+      return [rawValue] as string[] | number[] | boolean[];
```
Here, you also have to differentiate between `ArrayContains` and `ArrayContainsStrings`. `ArrayContains` has to potentially cast the value to a boolean or number (like the code below does), while `ArrayContainsStrings` can directly return the `rawValue`.

Valerionn reviewed (changes requested):
First of all: thank you for your PR!

I added prettier to this package and also applied it to your PR so that the PR is actually comparable and not 99% code style changes. For further commits, please just run `npm run prettier:write` in both projects (until I can add a pre-commit-hook or something similar).

The PR looks good so far - however, URL query parameters are always strings, and you currently treat them as strings/numbers/... (which is partially my fault for not typing these values correctly because I also only discovered this later...). 

_Usually_, Prisma does not like it if we give it strings when it expects numbers. It might be different when using arrays (since they're not typed) compared to the existing filters.

Have you tested this change in an actual Prisma project? If so, and it works, then we can merge the PR without my suggested changes. Otherwise, it would be great if you could please incorporate my comments so that we can be sure that it works.

memicmiralem committed:
fix string parsing on array filters

memicmiralem commented:
Hi @Valerionn 
thank you for your fast response!

I fixed now the string parsing issues, but without adding special `ArrayContainsStrings` for strings only. Created a helper fn `parseRawStringValue` for parsing which could probably also be reused on other places but I didn't want to mess with that now. As you can see the tests are now updated and passing, so it should be ok.

I haven't tested the new updates on the NestJS side (DirectFilterPipe, query -> findOptions object), but I did test all the filters on Prisma side if thats something you are interested in:

prisma.schema
```
model Computer {
  id          Int   @id @default(autoincrement())
  numberTags  Json? @default("[]")
  stringTags  Json? @default("[]")
  booleanTags Json? @default("[]")
}
```

index.ts
```
  const allComputers = await prisma.computer.findMany();
  console.log("🚀 all computers", allComputers);

  //ARRAY CONTAINS
  console.log("\narray_contains filters\n");

  const acFilter1 = await prisma.computer.findMany({
    where: { numberTags: { array_contains: [1, 2] } },
  });
  console.log("🚀 numberTags contains [1, 2]:", acFilter1);

  const acFilter2 = await prisma.computer.findMany({
    where: { stringTags: { array_contains: ["val3", "val4"] } },
  });
  console.log(`🚀 stringTags contains ["val3", "val4"]:`, acFilter2);

  const acFilter3 = await prisma.computer.findMany({
    where: { booleanTags: { array_contains: [true, false] } },
  });
  console.log("🚀 booleanTags contains [true, false]:", acFilter3);

  //ARRAY STARTS WITH
  console.log("\narray_starts_with filters\n");

  const aswFilter1 = await prisma.computer.findMany({
    where: { numberTags: { array_starts_with: 3 } },
  });
  console.log("🚀 numberTags starts with 3:", aswFilter1);

  const aswFilter2 = await prisma.computer.findMany({
    where: { stringTags: { array_starts_with: "val6" } },
  });
  console.log(`🚀 stringTags starts with "val6":`, aswFilter2);

  const aswFilter3 = await prisma.computer.findMany({
    where: { booleanTags: { array_starts_with: false } },
  });
  console.log("🚀 booleanTags starts with false:", aswFilter3);

  //ARRAY ENDS WITH
  console.log("\narray_ends_with filters\n");

  const aewFilter1 = await prisma.computer.findMany({
    where: { numberTags: { array_ends_with: 1 } },
  });
  console.log("🚀 numberTags ends with 1:", aewFilter1);

  const aewFilter2 = await prisma.computer.findMany({
    where: { stringTags: { array_ends_with: "val3" } },
  });
  console.log(`🚀 stringTags ends with "val3":`, aewFilter2);

  const aewFilter3 = await prisma.computer.findMany({
    where: { booleanTags: { array_ends_with: false } },
  });
  console.log("🚀 booleanTags ends with false:", aewFilter3);
```

console output
```
🚀 all computers [
  {
    id: 1,
    numberTags: [ 1, 2, 3 ],
    stringTags: [ 'val1', 'val2', 'val3' ],
    booleanTags: [ true, false, true ]
  },
  {
    id: 2,
    numberTags: [ 3, 4, 5 ],
    stringTags: [ 'val3', 'val4', 'val5' ],
    booleanTags: [ false, false, false ]
  },
  {
    id: 3,
    numberTags: [ 6, 7, 1 ],
    stringTags: [ 'val6', 'val7', 'val8' ],
    booleanTags: [ true, true ]
  }
]

array_contains filters

🚀 numberTags contains [1, 2]: [
  {
    id: 1,
    numberTags: [ 1, 2, 3 ],
    stringTags: [ 'val1', 'val2', 'val3' ],
    booleanTags: [ true, false, true ]
  }
]
🚀 stringTags contains ["val3", "val4"]: [
  {
    id: 2,
    numberTags: [ 3, 4, 5 ],
    stringTags: [ 'val3', 'val4', 'val5' ],
    booleanTags: [ false, false, false ]
  }
]
🚀 booleanTags contains [true, false]: [
  {
    id: 1,
    numberTags: [ 1, 2, 3 ],
    stringTags: [ 'val1', 'val2', 'val3' ],
    booleanTags: [ true, false, true ]
  }
]

array_starts_with filters

🚀 numberTags starts with 3: [
  {
    id: 2,
    numberTags: [ 3, 4, 5 ],
    stringTags: [ 'val3', 'val4', 'val5' ],
    booleanTags: [ false, false, false ]
  }
]
🚀 stringTags starts with "val6": [
  {
    id: 3,
    numberTags: [ 6, 7, 1 ],
    stringTags: [ 'val6', 'val7', 'val8' ],
    booleanTags: [ true, true ]
  }
]
🚀 booleanTags starts with false: [
  {
    id: 2,
    numberTags: [ 3, 4, 5 ],
    stringTags: [ 'val3', 'val4', 'val5' ],
    booleanTags: [ false, false, false ]
  }
]

array_ends_with filters

🚀 numberTags ends with 1: [
  {
    id: 3,
    numberTags: [ 6, 7, 1 ],
    stringTags: [ 'val6', 'val7', 'val8' ],
    booleanTags: [ true, true ]
  }
]
🚀 stringTags ends with "val3": [
  {
    id: 1,
    numberTags: [ 1, 2, 3 ],
    stringTags: [ 'val1', 'val2', 'val3' ],
    booleanTags: [ true, false, true ]
  }
]
🚀 booleanTags ends with false: [
  {
    id: 2,
    numberTags: [ 3, 4, 5 ],
    stringTags: [ 'val3', 'val4', 'val5' ],
    booleanTags: [ false, false, false ]
  }
]
```

Valerionn reviewed (approved):
Thanks again for your contribution and your tests! Your PR changes look good to me - the `parseRawStringHelper` is a great addition, I'll probably change the rest of the code in the future to use it as well.

I'll merge the PR, test it in our projects to make sure this change doesn't break anything (it shouldn't, but better to be safe than sorry) - and if everything works, I should be able to release a new version later this week.

memicmiralem commented:
@Valerionn when could we expect the new version to be released?
If there are any issues I can provide a sample repo with nestjs, prisma and these new changes

Valerionn commented:
@memicmiralem I've just released version 4.0.0 with your changes!
